src/common/variant8.cpp

```diff
@@ -0,0 +1,660 @@
+        if (n > 0)
+            switch (type) {
+            case VARIANT8_I8:
+                if ((i >= -128) && (i <= 127))
+                    var8.i8 = (int8_t)i;
+                else
+                    var8 = variant8_error(VARIANT8_ERR_OOFRNG, 0, 0);
+                break;
+            case VARIANT8_I16:
+                if ((i >= -32768) && (i <= 32767))
+                    var8.i16 = (int16_t)i;
+                else
+                    var8 = variant8_error(VARIANT8_ERR_OOFRNG, 0, 0);
+                break;
+            case VARIANT8_I32:
+                var8.i32 = (int32_t)i;
+                break;
+            }
+        break;
+    case VARIANT8_UI8:
+    case VARIANT8_UI16:
+    case VARIANT8_UI32:
+        n = sscanf(str, fmt ? fmt : "%u", &ui);
+        if (n > 0)
+            switch (type) {
+            case VARIANT8_UI8:
+                if (ui <= 255)
+                    var8.ui8 = (uint8_t)ui;
+                else
+                    var8 = variant8_error(VARIANT8_ERR_OOFRNG, 0, 0);
+                break;
+            case VARIANT8_UI16:
+                if (ui <= 65535)
+                    var8.ui16 = (uint16_t)ui;
+                else
+                    var8 = variant8_error(VARIANT8_ERR_OOFRNG, 0, 0);
+                break;
+            case VARIANT8_UI32:
+                var8.ui32 = (uint32_t)ui;
+                break;
+            }
+        break;
+    case VARIANT8_FLT:
+        n = sscanf(str, fmt ? fmt : "%f", &(var8.flt));
+        break;
+    case VARIANT8_CHAR:
+        n = sscanf(str, fmt ? fmt : "%c", &(var8.ch));
+        break;
+    case VARIANT8_USER:
+        n = sscanf(str, fmt ? fmt : "%u %hu %u", (unsigned int *)&(var8.usr32), (unsigned short int *)&(var8.usr16), &ui);
+        if (n >= 3)
+            var8.usr8 = ui;
+        break;
+    case VARIANT8_PCHAR:
+        var8 = variant8_pchar(0, strlen(str) + 1, 1);
+        n = sscanf(str, fmt ? fmt : "%s", var8.pch);
+        break;
+    default:
+        var8.type = VARIANT8_ERROR;
+        var8.err32 = VARIANT8_ERR_UNSCON;
+        break;
+    }
+    if (n <= 0) {
+        variant8_done(&var8);
+        var8.type = VARIANT8_ERROR;
+        var8.err32 = VARIANT8_ERR_INVFMT;
+    }
+    return var8;
+}
+
+uint32_t variant8_total_malloc_size = 0;
+
+void *variant8_malloc(uint16_t size) {
+#ifdef VARIANT8_DBG_MALLOC
+    uint8_t *ptr = (uint8_t *)(size ? malloc(size + sizeof(uint16_t)) : 0);
+    if (ptr) {
+        *((uint16_t *)ptr) = size;
+        variant8_total_malloc_size += size;
+    }
+    return ptr ? (ptr + sizeof(uint16_t)) : 0;
+#else  //VARIANT8_DBG_MALLOC
+    return malloc(size);
+#endif //VARIANT8_DBG_MALLOC
+}
+
+void variant8_free(void *ptr) {
+#ifdef VARIANT8_DBG_MALLOC
+    if (ptr) {
+        uint16_t size = ((uint16_t *)ptr)[-1];
+        variant8_total_malloc_size -= size;
+        free(((uint16_t *)ptr) - 1);
+    }
+#else  //VARIANT8_DBG_MALLOC
+    free(ptr);
+#endif //VARIANT8_DBG_MALLOC
+}
```

Comment:
Can be simplified by using `malloc_usable_size`

Reply:
i choosed this solution because malloc_usable_size is not supported in all compilers. I have some testing code under mingw... i will hack this different way.
commit 3074321